JDK-8290063: IGV: Give the graphs a unique number in the outline

tobiasholenstein opened the pull request:
Some graphs may have the same name more than once in IGV. To make it clearer which graph is currently open, all graphs within a group are now enumerated with `1.`, `2.`, `3.` , etc. Similarly, groups are enumerated with `1 -`, `2 -`, etc.

<img width="696" alt="overview" src="https://user-images.githubusercontent.com/71546117/198584142-887d323b-214d-4034-9b6f-318c697c8bdd.png">

The make it even further easier to distinguish different graphs and group, we can now rename them in two ways. 
**(A)** To rename a group or graph:
1. click on a graph or group once to select it (does not have to be opened). 
2. click a second time on the selected graph and wait 1-2 seconds. 
3. now you can rename the graph
4. on Linux F2 works as a shortcut for renaming
<img width="274" alt="rename_group" src="https://user-images.githubusercontent.com/71546117/198584168-2c3032fc-550f-42a9-9636-f0bf9483d18b.png">
<img width="311" alt="rename_graph" src="https://user-images.githubusercontent.com/71546117/198584177-dce038b8-ffed-411c-9547-c8f04b97474c.png">

**(B)** To rename a group or graph:
1. right-click on a graph or group once to select it (does not have to be opened). 
2. select "Rename.."
3. now a window opens to rename the graph
4. The shortcut for this `RenameAction` is `ALT-CTRL-R` or `OPT-CMD-R` on a mac keyboard
<img width="374" alt="RenameAction" src="https://user-images.githubusercontent.com/71546117/199456619-8b230fc1-4a4f-4ae1-b0a6-8e1540fe4372.png">
<img width="471" alt="Rename" src="https://user-images.githubusercontent.com/71546117/199456666-61b60849-0d08-40a5-a9a5-e4686ef7e587.png">

The numbering always starts at 1 and is continuous from 1 to N for N graphs. When a graph is deleted, the numbering of the following graphs changes. This implementation allows the keep the XML format unchanged, because the numbering is only local and not part of the name. However, if a graph/group is renamed, the name in the XML file will also change when it is saved.    

# Implementation 
The renaming is simply enabled by overriding `canRename() {return true;}` in `FolderNode` and `GraphNode`

The numbering is implemented in `getDisplayName()` in `Group` and `InputGraph` by concatenating the index of the group/graph with the name. 

When a group/graph is deleted the we trigger an update of the index in `FolderNode` -> `destroyNodes(Node[] nodes)` by calling `node.setDisplayName(node.getDisplayName())` for all nodes. 

Refresh the group/graph name in the `EditorTopComponent` of the opened graph is a bit tricky. It is implemented by adding a `Listener` to the `getDisplayNameChangedEvent()` of the currently opened `InputGraph` in `DiagramViewModel`. `getDisplayNameChangedEvent()` is fired whenever the name or the group name of the corresponding `InputGraph` is changed.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290063](https://bugs.openjdk.org/browse/JDK-8290063): IGV: Give the graphs a unique number in the outline


### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [4f36c623](https://git.openjdk.org/jdk/pull/10873/files/4f36c623178ac3bb2c30805f05f310d7cc361847)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [4f36c623](https://git.openjdk.org/jdk/pull/10873/files/4f36c623178ac3bb2c30805f05f310d7cc361847)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10873/head:pull/10873` \
`$ git checkout pull/10873`

Update a local copy of the PR: \
`$ git checkout pull/10873` \
`$ git pull https://git.openjdk.org/jdk pull/10873/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10873`

View PR using the GUI difftool: \
`$ git pr show -t 10873`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10873.diff">https://git.openjdk.org/jdk/pull/10873.diff</a>

</details>

TobiHartmann commented:
It works well for me but I'm wondering if "click and wait for 1-2 seconds" (which is more like 3 seconds on my system) should be replaced by an entry in the right-click menu. For fast renaming, a shortcut (F2) can be used, which already works.

I spotted the following (existing) issue:
- Open Phase 5
- Delete Phase 2
- The graph changes. Phase 6 is now displayed while the selection still shows Phase 5.

I don't think deleting a non-opened phase should affect the currently opened one, right?

tobiasholenstein commented:
> It works well for me but I'm wondering if "click and wait for 1-2 seconds" (which is more like 3 seconds on my system) should be replaced by an entry in the right-click menu. For fast renaming, a shortcut (F2) can be used, which already works.
> 
> I spotted the following (existing) issue:
> 
> * Open Phase 5
> * Delete Phase 2
> * The graph changes. Phase 6 is now displayed while the selection still shows Phase 5.
> 
> I don't think deleting a non-opened phase should affect the currently opened one, right?

Thanks for testing @TobiHartmann !

I agree that a right-click menu is more convenient. I added this functionality and updated the PR. 

The issue with the deleting is known and it already exists an bug report: https://bugs.openjdk.org/browse/JDK-8294066

TobiHartmann reviewed (approved):
Thanks for updating. Looks good to me!

chhagedorn commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/FolderNode.java:
```diff
@@ -141,6 +144,27 @@ public void setSelected(boolean selected) {
+    public void setName(String name) {
+        children.getFolder().setName(name);
+        fireDisplayNameChange(null, null);
+    }
```
Missing new line

tobiasholenstein replied:
fixed

chhagedorn reviewed (approved):
Nice improvement! Works as expected - looks good to me.

tobiasholenstein commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/GraphNode.java:
```diff
@@ -44,27 +45,49 @@
+    public void setName(String name) {
+        graph.setName(name);
+        fireDisplayNameChange(null, null);
+    }
```
fixed

tobiasholenstein commented:
Thanks @chhagedorn and @TobiHartmann for the reviews!

/integrate

tobiasholenstein commented:
/integrate
